File: go/interop_test.go
package securekit

// Cross-language interop tests. These load the shared fixtures in
// ../vectors/*.json (generated once from each language's own implementation)
// and verify this package can consume tokens/hashes produced by every other
// port. This is what actually proves the wire formats documented in the
// root README are portable, rather than just asserted by hand.

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"testing"
)

type passwordHashVectors struct {
	Password string            `json:"password"`
	Hashes   map[string]string `json:"hashes"`
}

type csrfTokenVectors struct {
	SecretBase64 string            `json:"secretBase64"`
	SessionID    string            `json:"sessionId"`
	Tokens       map[string]string `json:"tokens"`
}

type aeadVectors struct {
	KeyBase64           string              `json:"keyBase64"`
	Plaintext           string              `json:"plaintext"`
	Ciphertexts         map[string]string   `json:"ciphertexts"`
	CompatibilityGroups map[string][]string `json:"compatibilityGroups"`
}

func loadVectors(t *testing.T, name string, v interface{}) {
	t.Helper()
	data, err := os.ReadFile("../vectors/" + name)
	if err != nil {
		t.Fatalf("reading vector file %s: %v", name, err)
	}
	if err := json.Unmarshal(data, v); err != nil {
		t.Fatalf("parsing vector file %s: %v", name, err)
	}
}

func TestInteropPasswordHashes(t *testing.T) {
	var vecs passwordHashVectors
	loadVectors(t, "password_hashes.json", &vecs)

	hasher := NewDefaultPasswordHasher()
	for lang, hash := range vecs.Hashes {
		ok, err := hasher.Verify(vecs.Password, hash)
		if err != nil {
			t.Errorf("%s: unexpected error: %v", lang, err)
			continue
		}
		if !ok {
			t.Errorf("%s: hash did not verify against shared password (PHC format not portable)", lang)
		}
	}
}

func TestInteropCsrfTokens(t *testing.T) {
	var vecs csrfTokenVectors
	loadVectors(t, "csrf_tokens.json", &vecs)

	secret, err := base64.StdEncoding.DecodeString(vecs.SecretBase64)
	if err != nil {
		t.Fatalf("decoding secret: %v", err)
	}
	mgr := NewCsrfTokenManager(secret, 0) // ttl<=0: never expires, fixture-safe

	for lang, token := range vecs.Tokens {
		if !mgr.Verify(vecs.SessionID, token) {
			t.Errorf("%s: token did not verify (binary token format not portable)", lang)
		}
	}
}

func TestInteropAeadWithinCompatibilityGroup(t *testing.T) {
	var vecs aeadVectors
	loadVectors(t, "aead_ciphertexts.json", &vecs)

	key, err := base64.StdEncoding.DecodeString(vecs.KeyBase64)
	if err != nil {
		t.Fatalf("decoding key: %v", err)
	}
	enc := NewSymmetricEncryptor()

	// Go is documented to be in the "xchacha20poly1305" group. Every member
	// of that group must decrypt successfully; every language outside it
	// must fail -- this is the documented incompatibility, not a bug to
	// silently tolerate.
	group := vecs.CompatibilityGroups["xchacha20poly1305"]
	inGroup := make(map[string]bool)
	for _, lang := range group {
		inGroup[lang] = true
	}

	for lang, ctB64 := range vecs.Ciphertexts {
		ct, err := base64.StdEncoding.DecodeString(ctB64)
		if err != nil {
			t.Fatalf("%s: decoding ciphertext: %v", lang, err)
		}
		plaintext, err := enc.Decrypt(key, ct, nil)
		if inGroup[lang] {
			if err != nil {
				t.Errorf("%s: expected successful decrypt (same compatibility group), got %v", lang, err)
			} else if string(plaintext) != vecs.Plaintext {
				t.Errorf("%s: decrypted plaintext mismatch: got %q", lang, plaintext)
			}
		} else if err == nil {
			t.Errorf("%s: expected decrypt to fail (different AEAD construction), but it succeeded", lang)
		}
	}
}

File: php/tests/InteropTest.php
<?php

declare(strict_types=1);

namespace SecureKit\Tests;

use PHPUnit\Framework\TestCase;
use SecureKit\CsrfTokenManager;
use SecureKit\PasswordHasher;
use SecureKit\SymmetricEncryptor;

/**
 * Cross-language interop tests. Loads the shared fixtures in ../../vectors/*.json
 * (generated once from each language's own implementation) and verifies this
 * package can consume tokens/hashes/ciphertexts produced by every other port.
 */
final class InteropTest extends TestCase
{
    private const VECTORS_DIR = __DIR__ . '/../../vectors';

    private function loadVectors(string $name): array
    {
        $path = self::VECTORS_DIR . '/' . $name;
        $data = json_decode((string) file_get_contents($path), true, 512, JSON_THROW_ON_ERROR);
        return $data;
    }

    public function testPasswordHashesInterop(): void
    {
        $vecs = $this->loadVectors('password_hashes.json');
        $hasher = new PasswordHasher();

        foreach ($vecs['hashes'] as $lang => $hash) {
            $this->assertTrue(
                $hasher->verify($vecs['password'], $hash),
                "$lang: hash did not verify against shared password (PHC format not portable)"
            );
        }
    }

    public function testCsrfTokensInterop(): void
    {
        $vecs = $this->loadVectors('csrf_tokens.json');
        $secret = base64_decode($vecs['secretBase64'], true);
        $mgr = new CsrfTokenManager($secret, 0); // ttl<=0: never expires, fixture-safe

        foreach ($vecs['tokens'] as $lang => $token) {
            $this->assertTrue(
                $mgr->verify($vecs['sessionId'], $token),
                "$lang: token did not verify (binary token format not portable)"
            );
        }
    }

    public function testAeadInteropWithinCompatibilityGroup(): void
    {
        $vecs = $this->loadVectors('aead_ciphertexts.json');
        $key = base64_decode($vecs['keyBase64'], true);
        $enc = new SymmetricEncryptor();

        $group = array_flip($vecs['compatibilityGroups']['xchacha20poly1305']);

        foreach ($vecs['ciphertexts'] as $lang => $ctB64) {
            $ct = base64_decode($ctB64, true);
            if (isset($group[$lang])) {
                $plaintext = $enc->decrypt($key, $ct);
                $this->assertSame($vecs['plaintext'], $plaintext, "$lang: decrypted plaintext mismatch");
            } else {
                try {
                    $enc->decrypt($key, $ct);
                    $this->fail("$lang: expected decrypt to fail (different AEAD construction), but it succeeded");
                } catch (\RuntimeException $e) {
                    $this->assertSame('securekit: decryption failed', $e->getMessage());
                }
            }
        }
    }
}
